docs/source/more_types.rst

```diff
@@ -178,6 +183,386 @@ create subclasses of these objects either.
+certainly a mistake, mypy will report an error. To fix the error, we can
+either 1) delete the second overload or 2) swap the order of the overloads:
+
+.. code-block:: python
+
+    # Everything is ok now -- the variants are correctly ordered
+    # from most to least specific.
+
+    @overload
+    def add(left: Literal, right: Literal) -> Literal: ...
+
+    @overload
+    def add(left: Expression, right: Expression) -> Expression: ...
+
+    def add(left: Expression, right: Expression) -> Expression:
+        # ...snip...
+
+Mypy will also type check the different variants and flag any overloads
+that have inherently unsafely overlapping variants. For example, consider
+the following unsafe overload definition:
+
+.. code-block:: python
+
+    from typing import overload, Union
+
+    @overload
+    def unsafe_func(x: int) -> int: ...
+
+    @overload
+    def unsafe_func(x: object) -> str: ...
+
+    def unsafe_func(x: object) -> Union[int, str]:
+        if isinstance(x, int):
+            return 42
+        else:
+            return "some string"
+
+On the surface, this function definition appears to be fine. However, it will
+result in a discrepency between the inferred type and the actual runtime type
+when we try using it like so:
+
+.. code-block:: python
+
+    some_obj: object = 42
+    unsafe_func(some_obj) + " danger danger"  # Type checks, yet crashes at runtime!
+
+This program type checks according to the annotations, but will actually crash
+at runtime! Since ``some_obj`` is of type ``object``, mypy will decide that
+``unsafe_func`` must return something of type ``str`` and concludes the above will
+type check. But in reality, ``unsafe_func`` will return an int, causing the code
+to crash at runtime!
+
+To prevent these kinds of issues, mypy will detect and prohibit inherently unsafely
+overlapping overloads on a best-effort basis. Two variants are considered unsafely
+overlapping when both of the following are true:
+
+1. All of the arguments of the first variant are compatible with the second.
+2. The return type of the first variant is *not* compatible with (e.g. is not a
+   subtype of) the second.
+
+So in this example, the ``int`` argument in the first variant is a subtype of
+the ``object`` argument in the second, yet the ``int`` return type not is a subtype of
+``str``. Both conditions are true, so mypy will correctly flag ``unsafe_func`` as
+being unsafe.
+
+However, mypy will not detect *all* unsafe uses of overloads. For example,
+suppose we modify the above snippet so it calls ``summarize`` instead of
+``unsafe_func``:
+
+.. code-block:: python
+
+    some_list: List[str] = []
+    summarize(some_list) + "danger danger"  # Type safe, yet crashes at runtime!
+
+We run into a similar issue here. This program type checks if we look just at the
+annotations on the overloads. But since ``summarize(...)`` is designed to be biased
+towards returning an int when it receives an empty list, this program will actually
```

Comment:
I would format this as ` ``int`` ` (or rather ` ``float`` ` if you update the signature as suggested).